walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -13,48 +13,84 @@
 // limitations under the License.
 package com.google.sps;
 
+import java.util.Arrays;
+import java.util.ArrayList;
 import java.util.Collection;
 import java.util.Collections;
-import java.util.ArrayList;
+import java.util.HashSet;
 import java.util.List;
 
+
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    List<TimeRange> availableTimes = new ArrayList<TimeRange>();
-    List<TimeRange> unavailableTimes = new ArrayList<TimeRange>();
-
+    List<TimeRange> availableTimesForMandatory = new ArrayList<TimeRange>();
+    List<TimeRange> availableTimesForOptional = new ArrayList<TimeRange>();
+    List<TimeRange> unavailableTimesForMandatory = new ArrayList<TimeRange>();
+    List<TimeRange> unavailableTimesForOptional = new ArrayList<TimeRange>();
+    
     for (Event e : events) {
-        if (!attendeesFree(e, request)) {
-            unavailableTimes.add(e.getWhen());
+        if (!mandatoryAttendeesFree(e, request)) {
+            unavailableTimesForMandatory.add(e.getWhen());
+        }
+
+        if (!optionalAttendeesFree(e, request)) {
+            unavailableTimesForOptional.add(e.getWhen());
         }
     }
 
-    Collections.sort(unavailableTimes, TimeRange.ORDER_BY_START);
+    Collections.sort(unavailableTimesForMandatory, TimeRange.ORDER_BY_START);
+    Collections.sort(unavailableTimesForOptional, TimeRange.ORDER_BY_START);
 
-    if (unavailableTimes.size() > 0) {
-        combineOverlaps(unavailableTimes, unavailableTimes.get(0), 0);
+    if (unavailableTimesForMandatory.size() > 0) { // Do this only for mandatory attendees since only one optional attendee exists
+        combineOverlaps(unavailableTimesForMandatory, unavailableTimesForMandatory.get(0), 0);
     }
 
-    int start = 0;
-    for (TimeRange range : unavailableTimes) {
-        TimeRange availableRange = TimeRange.fromStartEnd(start, range.start(), false);
+    findAvailableRanges(unavailableTimesForMandatory, availableTimesForMandatory, request);
+    findAvailableRanges(unavailableTimesForOptional, availableTimesForOptional, request);
+
+    List<TimeRange> finalAvailableTimes = new ArrayList<TimeRange>();
+
+    if (request.getOptionalAttendees().size() == 0) {
+        finalAvailableTimes = availableTimesForMandatory;
+    } else if (request.getAttendees().size() == 0) {
+        finalAvailableTimes = availableTimesForOptional; 
+    } else if (availableTimesForMandatory.size() == 0) {
+        finalAvailableTimes = availableTimesForOptional;
+    } else if (availableTimesForOptional.size() == 0) { 
+        finalAvailableTimes = availableTimesForMandatory;
+    } else {
```

Comment:
You can group alike statements. For example:

```
if (request.getAttendees().size() == 0 || availableTimesForMandatory.size() == 0) {
    finalAvailableTimes = availableTimesForOptional;
} else if (request.getOptionalAttendees().size() == 0 || availableTimesForOptional.size() == 0) {
    finalAvailableTimes = availableTimesForMandatory;
}
```

Reply:
May need to modify the statements but generally, if the code blocks are doing the same thing, you should be able to collapse multiple if/else statements.

Reply:
Yeah I tried to group the statements since they executed the same task, but then I saw that the order of the if/else chain determined if the tests passed or not. For example, one test failed when I grouped the conditionals because the order of the if/else statements determined whether finalAvaialbleTimes was set to availableTimesForOptional or availableTimesForMandatory